### PR TITLE
Add new methods to Revit_Core_Engine

### LINKED ISSUE #1343: Additional methods for Revit_Core_Engine
#### Description:

To use Geometry_Engine methods, we often need to convert Revit native geometry elements to Bhom objects. This becomes inefficient in heavy applications. Therefore, we need a few additional geometry methods in Revit_Core_Engine to handle use cases like:
- Get the intersection points of 2 vectors
- Get the smallest possible angle between 2 vectors
- Get the closest point from a point cloud
- Check if a line or face is vertical

### PR DESCRIPTION
**Pull request by vietle-bh**
### Issues addressed by this PR

Closes #1343 

Added new methods such as:
- Get the intersection points of 2 vectors
- Get the smallest possible angle between 2 vectors
- Get the closest point from a point cloud
- Check if a line or face is vertical

### CODE REVIEW

**Review by pawelbaran — CHANGES_REQUESTED** [review]
The code does not compile...

Also I do not think you need the geometry methods, taken we have all of them in Geometry_Engine.

File: Revit_Core_Engine/Query/LocationPoint.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Gets the location point of a Revit element.")]
+        [Input("elem", "A generic Element instance.")]
+        [Input("useCurveMidPoint", "If True, get the middle point of the element's LocationCurve if the element isn't point-based.")]
+        [Output("xyz", "The location point of a Revit element.")]
+        public static XYZ LocationPoint(this Element elem, bool useCurveMidPoint = false)
```
[the remark this answers is not in the document]
Comment (michal-pekacki) [reply Revit_Core_Engine/Query/LocationPoint.cs:39]: I used this in Revit_Tools:
https://github.com/BuroHappoldEngineering/Revit_Tools/blob/46ea8bd0477f281468fc7cab79ef6adcf17fb580/Revit_Mechanical_Engine/Compute/ReferenceLevel.cs#L203-L229

[the remark this answers is not in the document]
Comment (vietle-bh) [reply Revit_Core_Engine/Query/LocationPoint.cs:39]: Michal's method can indeed do the same thing. I'll move it here.

File: Revit_Core_Engine/Compute/VectorIntersection.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Gets the intersection point of 2 vectors given their start/end points and directions.")]
+        [Input("startPoint", "A point on the first vector.")]
+        [Input("endPoint", "A point on the second vector.")]
+        [Input("startDir", "The direction of the first vector.")]
+        [Input("endDir", "The direction of the second vector.")]
+        [Output("xyz", "The intersection point of 2 vectors given their start/end points and directions.")]
+        public static XYZ VectorIntersection(this XYZ startPoint, XYZ endPoint, XYZ startDir, XYZ endDir)
```
[the remark this answers is not in the document]
Comment (vietle-bh) [reply Revit_Core_Engine/Compute/VectorIntersection.cs:43]: If I replaced this with LineIntersection, this:
`XYZ elbowPnt = Core.Compute.VectorIntersection(startPoint, endPoint, startDir, endDir);`

will turn into:

```c#
var line1 = new BhLine { Start = startPoint.PointFromRevit(), End = (startPoint + startDir).PointFromRevit() };
var line2 = new BhLine { Start = endPoint.PointFromRevit(), End = (endPoint + endDir).PointFromRevit() };

line1.Infinite = true;
line2.Infinite = true;

XYZ elbowPnt = line1.LineIntersection(line2, true).ToRevit();
```

Is it worth it?

[the remark this answers is not in the document]
Comment (pawelbaran) [reply Revit_Core_Engine/Compute/VectorIntersection.cs:43]: I think so, but I would first ask if you will still need it if all geometry/maths will be done on the BHoM side, not on Revit?

[the remark this answers is not in the document]
Comment (vietle-bh) [reply Revit_Core_Engine/Compute/VectorIntersection.cs:43]: Sure, I'll change to using the longer form.
The example above is in Revit_Toolkit where my new methods purely deal with Revit objects.

[the remark this answers is not in the document]
Comment (vietle-bh) [reply Revit_Core_Engine/Compute/VectorIntersection.cs:43]: So geometry/math calls are on the Revit side.

**Comment by vietle-bh** [comment]
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check null-handling
@BHoMBot check serialisation

**Comment by vietle-bh** [comment]
Closing to reopen this PR because the commit history has rewritten itself.

**Pull request closed without merging**